component/sqs/component.go

```diff
@@ -0,0 +1,332 @@
+		},
+		[]string{"state"},
+	)
+	prometheus.MustRegister(queueSize)
+}
+
+// Component implementation of a async component.
+type Component struct {
+	name              string
+	queueName         string
+	queueURL          string
+	sqsAPI            sqsiface.SQSAPI
+	maxMessages       *int64
+	pollWaitSeconds   *int64
+	visibilityTimeout *int64
+	statsInterval     time.Duration
+	proc              ProcessorFunc
+	retries           uint
+	retryWait         time.Duration
+}
+
+// New creates a new component with support for functional configuration.
+func New(name, queueName string, sqsAPI sqsiface.SQSAPI, proc ProcessorFunc, oo ...OptionFunc) (*Component, error) {
+	if name == "" {
+		return nil, errors.New("component name is empty")
+	}
+
+	if queueName == "" {
+		return nil, errors.New("queue name is empty")
+	}
+
+	if sqsAPI == nil {
+		return nil, errors.New("SQS API is nil")
+	}
+
+	if proc == nil {
+		return nil, errors.New("process function is nil")
+	}
+
+	out, err := sqsAPI.GetQueueUrlWithContext(context.Background(), &sqs.GetQueueUrlInput{
+		QueueName: aws.String(queueName),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get queue URL: %w", err)
+	}
+
+	cmp := &Component{
+		name:              name,
+		queueName:         queueName,
+		queueURL:          aws.StringValue(out.QueueUrl),
+		sqsAPI:            sqsAPI,
+		maxMessages:       aws.Int64(3),
+		pollWaitSeconds:   nil,
+		visibilityTimeout: nil,
+		statsInterval:     10 * time.Second,
+		proc:              proc,
+		retries:           10,
+		retryWait:         time.Second,
+	}
+
+	for _, optionFunc := range oo {
+		err = optionFunc(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Run starts the consumer processing loop messages.
+func (c *Component) Run(ctx context.Context) error {
+	chErr := make(chan error)
+
+	go c.consume(ctx, chErr)
+
+	tickerStats := time.NewTicker(c.statsInterval)
+	defer tickerStats.Stop()
+	for {
+		select {
+		case err := <-chErr:
+			return err
+		case <-ctx.Done():
+			return nil
+		case <-tickerStats.C:
+			err := c.report(ctx, c.sqsAPI, c.queueURL)
+			if err != nil {
+				log.FromContext(ctx).Errorf("failed to report sqsAPI stats: %v", err)
+			}
+		}
+	}
+}
+
+func (c *Component) consume(ctx context.Context, chErr chan error) {
+	logger := log.FromContext(ctx)
+
+	retries := c.retries
+
+	for {
+		if ctx.Err() != nil {
+			return
+		}
+		logger.Debugf("consume: polling SQS sqsAPI %s for %d messages", c.queueName, *c.maxMessages)
+		output, err := c.sqsAPI.ReceiveMessageWithContext(ctx, &sqs.ReceiveMessageInput{
+			QueueUrl:            &c.queueURL,
+			MaxNumberOfMessages: c.maxMessages,
+			WaitTimeSeconds:     c.pollWaitSeconds,
+			VisibilityTimeout:   c.visibilityTimeout,
+			AttributeNames: aws.StringSlice([]string{
+				sqsAttributeSentTimestamp,
+			}),
+			MessageAttributeNames: aws.StringSlice([]string{
+				sqsMessageAttributeAll,
+			}),
+		})
+		if err != nil {
```

Comment:
i am wondering if it would make sense to abstract the failure strategy logic, so that it s reusable as a component ... and also make this method size smaller ..
i.e. 
```
if err != nil {
    ok := handleRetry(...)
    if !ok {
        return
    }
}
```

Reply:
I would avoid introducing a function here since it is just 5 lines of code and the funtion would be 3 lines which makes it not that much of a benefit. as for reusability maybe we can see in the future how the other native components work out and introduce then something that is common. What do you think?